src/Workspaces/Remote/Core/Diagnostics/PerformanceQueue.cs

```diff
@@ -0,0 +1,226 @@
+                _snapshots.AddLast(first);
+            }
+        }
+
+        public void GetPerformanceData(Dictionary<string, (double mean, double stddev)> aggregatedPerformanceDataPerAnalyzer)
+        {
+            if (_snapshots.Count < MinSampleSize)
+            {
+                // we don't have enough data to report this
+                return;
+            }
+
+            using (var pooledMap = SharedPools.Default<Dictionary<int, string>>().GetPooledObject())
+            using (var pooledSet = SharedPools.Default<HashSet<int>>().GetPooledObject())
+            using (var pooledList = SharedPools.Default<List<double>>().GetPooledObject())
+            {
+                var reverseMap = pooledMap.Object;
+                AnalyzerNumberAssigner.Instance.GetReverseMap(reverseMap);
+
+                var analyzerSet = pooledSet.Object;
+
+                // get all analyzers
+                foreach (var snapshot in _snapshots)
+                {
+                    snapshot.AppendAnalyzers(analyzerSet);
+                }
+
+                var list = pooledList.Object;
+
+                // calculate aggregated data per analyzer
+                foreach (var assignedAnalyzerNumber in analyzerSet)
+                {
+                    foreach (var snapshot in _snapshots)
+                    {
+                        var timeSpan = snapshot.GetTimeSpanInMillisecond(assignedAnalyzerNumber);
+                        if (timeSpan == null)
+                        {
+                            // not all snapshot contains all analyzers
+                            continue;
+                        }
+
+                        list.Add(timeSpan.Value);
+                    }
+
+                    // data is only stable once we have more than certain set
+                    // of samples
+                    if (list.Count < MinSampleSize)
+                    {
+                        continue;
+                    }
+
+                    // set performance data
+                    aggregatedPerformanceDataPerAnalyzer[reverseMap[assignedAnalyzerNumber]] = GetMeanAndAdjustedStandardDeviation(list);
+
+                    list.Clear();
+                }
+            }
+        }
+
+        private (double, double) GetMeanAndAdjustedStandardDeviation(List<double> data)
+        {
+            var mean = data.Average();
+            var stddev = Math.Sqrt(data.Select(ms => Math.Pow(ms - mean, 2)).Average());
+            var squareLength = Math.Sqrt(data.Count);
+
+            return (mean, stddev / squareLength);
+        }
+
+        private class Snapshot
+        {
+            /// <summary>
+            /// Raw performance data. 
+            /// Keyed by analyzer unique number got from AnalyzerNumberAssigner.
+            /// Value is delta (TimeSpan - minSpan) among span in this snapshot
+            /// </summary>
+            private readonly Dictionary<int, double> _performanceMap;
```

Comment:
:bulb: If `AnalyzerNumberAssigner` is sequential and has a known maximum, consider using `double[]`, where the index is the key from the map.

Reply:
it doesn't have known maximum and since analyzer can be added and removed dynamically, linear array might require more resizing.